packages/@aws-cdk/aws-s3-deployment/lib/source.ts

```diff
@@ -54,7 +56,7 @@ export class Source {
   public static asset(path: string): ISource {
     return {
       bind(context: cdk.Construct): SourceConfig {
-        const asset = new s3_assets.Asset(context, 'Asset', { path });
+        const asset = new s3_assets.Asset(context, `Asset${++nextUniqueId}`, { path });
```

Comment:
Sorry but this is not good enough. Just realizing the ID is global, which means that it can be mutated by things out of the user's control. For example, if suddenly some 3rd party construct library is changed and uses s3-deployment under the hood, the ID can be changed for all s3 deployments.

Let's find a way to make this "context-local". For example, you can use `context.node.tryFindChild` or you can allocate the ID outside and pass it to `bind`.